File: 20-valid-parentheses/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> LIFO;
        for(int i = 0; i<s.size(); i++){
            if(s[i]=='(' || s[i]=='[' || s[i]=='{'){
                LIFO.push(s[i]);
            } else{
                if(LIFO.empty()){
                    return false;
                } else if(s[i]==')'){
                    if(LIFO.top()=='('){
                        LIFO.pop();
                    } else{
                        return false;
                    }
                } else if(s[i]=='}'){
                    if(LIFO.top()=='{'){
                        LIFO.pop();
                    } else{
                        return false;
                    }
                } else if(s[i]==']'){
                    if(LIFO.top()=='['){
                        LIFO.pop();
                    } else{
                        return false;
                    }
                }
            }
        }
        if(LIFO.empty()){
            return true;
        } else{
            return false;
        }
    }
};

File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        seen = []
        closedToOpen = {"]":"[", "}":"{", ")":"("}

        for c in s:
            if c in closedToOpen:
                if seen and seen[-1]==closedToOpen[c]:
                    seen.pop()
                else:
                    return False
            else:
                seen.append(c)
        return True if not seen else False